### PR TITLE
Clean up the `instantiate` method

### PR DESCRIPTION
**Pull request by shagunsodhani**
## Motivation

Clean up the `instantiate` method

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

Add some new tests

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

#764

### CODE REVIEW

**Comment by shagunsodhani** [comment]
@omry you could check the new section in documentation

File: hydra/utils.py [hunk 1]
```diff
@@ -34,8 +36,10 @@ def call(config: Union[ObjectConf, DictConfig], *args: Any, **kwargs: Any) -> An
         else:
             assert callable(type_or_callable)
             return _call_callable(type_or_callable, config, *args, **kwargs)
+    except ValueError as e:
+        raise HydraException(f"Error calling function : {e}") from e
     except Exception as e:
-        raise HydraException(f"Error instantiating '{cls}' : {e}") from e
+        raise HydraException(f"Error calling '{cls}' : {e}") from e
```
Comment (omry) [inline hydra/utils.py:42]: Differentiating between the two errors based on these exceptions is very fragile.
I suggest that you put the exception re-raising in call and instantiate, and have them both call a common private function. This way you know if this is a call or instantiating based on the context.

File: website/docs/advanced/instantiate_objects.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+```python
+def instantiate(config: Union[ObjectConf, DictConfig], *args: Any, **kwargs: Any) -> Any:
+    """
+    :param config: An ObjectConf or DictConfig describing what to instantiate and what params to use
+    :param args: optional positional parameters pass-through
+    :param kwargs: optional named parameters pass-through
+    :return: the return value from the specified class
+    """
+```
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:27]: I think no point in documenting both.
include instantiate and mention that call is available as an alias.

File: website/docs/advanced/instantiate_objects.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
+---
+id: instantiate
+title: Instantiate objects and calling functions
+sidebar_label: Instantiate objects and calling functions
+---
+
+Hydra provides `hydra.utils.call()` (and its alias `hydra.utils.instantiate()`) for instantiating objects and calling functions While `instantiate` is an alias for `call`, you may prefer to use `call` for invoking functions and class methods, and `instantiate` for creating objects.
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:7]:
```suggestion
Hydra provides `hydra.utils.instantiate()` (and its alias `hydra.utils.call()`) for instantiating objects and calling functions Prefer `instantiate` for creating objects and `call` for invoking functions.
```

File: website/docs/advanced/instantiate_objects.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    :param config: An ObjectConf or DictConfig describing what to instantiate and what params to use
+    :param args: optional positional parameters pass-through
+    :param kwargs: optional named parameters pass-through
+    :return: the return value from the specified class
+    """
+```
+
+For using these functions, the config must have a key called `target`. Older versions of Hydra also supported the `cls` and `class` key but they are deprecated now (hence not recommended) and will be removed in Hydra 1.1 . If both `target` and `cls` (or `class`) keys are present, `target` key will be used.
+
+If a key called `params` is also present, its value is passed as keyword arguments to class/function specificed by `target`.
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:31]: Don't document the deprecated behavior. the deprecation warning should be enough.

File: website/docs/advanced/instantiate_objects.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+```python
+@dataclass
+class ObjectConf(Dict[str, Any]):
+    # class, class method or function name
+    target: str = MISSING
+    # parameters to pass to target when calling it
+    params: Any = field(default_factory=dict)
+```
+
+### Example config node for hydra.utils.call
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:44]: Nothing about this is about call.
Those are two examples, one using ObjectConf and one using yaml config.

Explain when each one should be used, and put the config one first (I think config usage is more common).

**Comment by omry** [comment]
I can tell that you did not test what this documentation looks like because it's missing the sidebar changes.
See the readme in website and visualize the website.

Also, rebase. The lint error is fixed on master.

File: website/docs/advanced/instantiate_objects.md (hunk 2 above)
Comment (shagunsodhani) [reply website/docs/advanced/instantiate_objects.md:27]: Sounds good. Probably it doesnt matter but earlier documentation always mentioned `call` (as the primary method) and `instantiate` as the alias (I know no real difference). I will stick to that convention.

File: hydra/utils.py (hunk 1 above)
Comment (shagunsodhani) [reply hydra/utils.py:42]: ohh just to give some context, these are just different kind of exceptions - missing `target` key in config, vs exception. I think its better if always raise a `HydraException` (and use it to wrap a `ValueError`). Should we make a new `HydraException` class (that extends `ValueError`) ?

**Comment by shagunsodhani** [comment]
`yamllint` should ignore some directories

**Comment by shagunsodhani** [comment]
I am not sure why this change https://github.com/facebookresearch/hydra/pull/766/commits/84feaea52fc29152a7821050ea677101b6345cc9 is suddenly needed (or needed only on windows/mac). I am installing Hydra locally to reproduce it.

**Comment by shagunsodhani** [comment]
Locally, the mac tests pass without making #84feaea change to `yamllint`. cc @omry

File: hydra/utils.py (hunk 1 above)
Comment (omry) [reply hydra/utils.py]: Who is throwing a ValueError on a missing target field?
You realize that ValueError is very common any instantiated code is likely to throw it too?

Comment (shagunsodhani) [reply hydra/utils.py]: Now we throw a HydraException if the `target` key is missing. Now might be a good time to have a `MissingKey` exception for Hydra, along the lines of `MissingConfig` exception. The `MissingKey` exception can be used by HPO-plugins as well.

Comment (omry) [reply hydra/utils.py]: OmegaConf got that, no need to replicate.
What happens if you just access target blindly?

Comment (shagunsodhani) [reply hydra/utils.py]: Used to raise a `ValueError`

Comment (omry) [reply hydra/utils.py]: > Who is throwing a ValueError on a missing target field?

Comment (shagunsodhani) [reply hydra/utils.py]: https://github.com/facebookresearch/hydra/blob/master/hydra/_internal/utils.py#L561 . Last changed in https://github.com/facebookresearch/hydra/commit/a32958846d488aa85d324b11a3f2ac23a0fcbda2 when deprecating `cls`.

Comment (omry) [reply hydra/utils.py]: and if we just make it a HydraException there?

Comment (shagunsodhani) [reply hydra/utils.py]: Yeah did that, based on your feedback: https://github.com/facebookresearch/hydra/pull/766/commits/54107b91349173058cd1c45d3200863dc43e5b23

File: website/docs/advanced/instantiate_objects.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  def class_method(self, z: int) -> Any:
+    return self(z, 10)
+
+  @staticmethod
+  def static_method(z: int) -> int:
+    return z + 1
+
+def bar(z: int) -> int:
+  return z + 2
+```
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:61]:
```suggestion
models.py
```python
class Foo:
  # target: models.Foo
  def __init__(self, x: int, y: int) -> None:
    self.x = x
    self.y = y

  # target: models.Foo.class_method
  @classmethod
  def class_method(self, z: int) -> Any:
    return self(z, 10)
    
  # target: models.Foo.static_method
  @staticmethod
  def static_method(z: int) -> int:
    return z + 1

# target: models.bar
def bar(z: int) -> int:
  return z + 2
```
```

File: website/docs/advanced/instantiate_objects.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+mystaticmethod:
+  target: models.Foo.static_method
+  params:
+    z: 15
+
+myfunction:
+  target: models.bar
+  params:
+    z: 15
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:83]:
```suggestion
For example, to instantiate a models.Foo object:
config.yaml
```yaml
myobject:
  target: models.Foo
  params:
    x: 10
    y: 20
```

File: website/docs/advanced/instantiate_objects.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+```python
+import hydra
+
+@hydra.main(config_path="config.yaml")
+def app(cfg):
+  foo1: Foo = hydra.utils.instantiate(cfg.myobject)  # Foo(10, 20)
+  foo2: Foo = hydra.utils.instantiate(cfg.myclassmethod)  # Foo(5, 10)
+  ret1: int = hydra.utils.instantiate(cfg.mystaticmethod)  # 16
+  ret2: int = hydra.utils.instantiate(cfg.myfunction)  # 17
+```
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:97]:
```suggestion
Now to test these, `instantiate` (or `call`) them as follows:

```python
# foo is as described in the config
foo : Foo = hydra.utils.instantiate(cfg.myobject)

# you can also override the config values on the callsite:
foo2 : Foo = hydra.utils.instantiate(cfg.myobject, x=100)

 and even pass additional fields that are not in the config. 
```

File: website/docs/advanced/instantiate_objects.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+These methods also support functions built into the standard library:
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:99]: simplify this one like the examples I put down so far. (probably fold it into the my example above).

File: website/docs/advanced/instantiate_objects.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+```python
+import hydra
+
+@hydra.main(config_path="config.yaml")
+def app(cfg):
+  foo: str = hydra.utils.instantiate(cfg)  # "42"
+
+```
+
+Another useful scenario is when we want to create the `None` object. In that case, set `target = None`:
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:117]: way too much details, this fits in a single sentence.
It's also wrong, it's not the target that should be None, it's the entire config node.

File: website/sidebars.js [hunk 11]
```diff
@@ -11,51 +11,51 @@ module.exports = {
...
-                    label: 'Running your Hydra app',
-                    items: [
-                        'tutorials/basic/running_your_app/multi-run',
-                        'tutorials/basic/running_your_app/working_directory',
-                        'tutorials/basic/running_your_app/logging',
-                        'tutorials/basic/running_your_app/debugging',
-                        'tutorials/basic/running_your_app/tab_completion',
-                    ]
-                },
+            'tutorials/intro',
```
Comment (omry) [inline website/sidebars.js:14]: Don't reformat.
if you want to reformat it, create a different issue and also mention how to achieve the same formatting.

File: website/docs/advanced/instantiate_objects.md (hunk 6 above)
Comment (shagunsodhani) [reply website/docs/advanced/instantiate_objects.md:61]: done

File: website/docs/advanced/instantiate_objects.md [hunk 12]
```diff
@@ -44,66 +44,58 @@ class ObjectConf(Dict[str, Any]):
 models.py
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:44]: rename to example.py (and update docs and config)

File: website/docs/advanced/instantiate_objects.md [hunk 13]
```diff
@@ -44,66 +44,58 @@ class ObjectConf(Dict[str, Any]):
...
 def bar(z: int) -> int:
   return z + 2
+
 ```
+
+For example, to instantiate a models.Foo object:
+
 config.yaml
 ```yaml
 myobject:
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:74]: rename myobject to foo, then instantiate becomes:

foo : Foo = instantiate(cfg.foo)

File: website/docs/advanced/instantiate_objects.md [hunk 14]
```diff
@@ -44,66 +44,58 @@ class ObjectConf(Dict[str, Any]):
...
-import hydra
-
-@hydra.main(config_path="config.yaml")
-def app(cfg):
-  foo1: Foo = hydra.utils.instantiate(cfg.myobject)  # Foo(10, 20)
-  foo2: Foo = hydra.utils.instantiate(cfg.myclassmethod)  # Foo(5, 10)
-  ret1: int = hydra.utils.instantiate(cfg.mystaticmethod)  # 16
-  ret2: int = hydra.utils.instantiate(cfg.myfunction)  # 17
+# foo is as described in the config
+foo : Foo = hydra.utils.instantiate(cfg.myobject)
```
Comment (omry) [inline website/docs/advanced/instantiate_objects.md:84]:
```
from hydra.utils import instantiate
```
on need for fully qualified name everywhere.

File: website/sidebars.js [hunk 15]
```diff
@@ -1,123 +1,123 @@
...
-            'fb/internal-fb-cluster',
-            'fb/fair-cluster',
-        ]),
-    }
-}
+    return process.env.FB_INTERNAL ? elements : [];
+ }
+ 
+ module.exports = {
+     docs: {
```
Comment (omry) [inline website/sidebars.js:8]: You are still formatting this file.

File: website/docs/patterns/objects.md [hunk 16]
```diff
@@ -3,96 +3,6 @@ id: objects
...
-
-@hydra.main(config_path="config.yaml")
-def app(cfg):
-  foo1: Foo = hydra.utils.call(cfg.myobject)  # Foo(10, 20)
-  foo2: Foo = hydra.utils.call(cfg.myclassmethod)  # Foo(5, 10)
-  ret1: int = hydra.utils.call(cfg.mystaticmethod)  # 16
-  ret2: int = hydra.utils.call(cfg.myfunction)  # 17
-```
-### Real World Example
 One of the best ways to drive different behavior in the application is to instantiate different implementations of an interface.
```
Comment (omry) [inline website/docs/patterns/objects.md:6]: Please make a critical pass over this doc to see that it still makes sense.

Comment (shagunsodhani) [reply website/docs/patterns/objects.md:6]: Yeah - sorry I thought I had marked the PR as draft - its not ready for review.

**Comment by shagunsodhani** [comment]
@omry you can take a look now. Other than the change of renaming "overview_of_instantiating.." -> "overview", other things should be fixed. Once you approve, I will rebase and (force) push again. Not doing it right now it will be a force push.

File: website/docs/patterns/instantiate_objects/overview.md [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+
+With this, you can instantiate the object from the configuration with a single line of code:
+```python
+@hydra.main(config_path="conf", config_name="config")
+def my_app(cfg):
+    connection = hydra.utils.instantiate(cfg.db)
+    connection.connect()
+```
+
+Learn more about instantiating objects and functions [here](instantiate_objects_reference).
```
Comment (omry) [inline website/docs/patterns/instantiate_objects/overview.md:88]: and calling functions.

maybe just remove this, it's in the next page anyway.

File: website/sidebars.js (hunk 15 above)
Comment (omry) [reply website/sidebars.js:8]: Can you not reformat this file?

**Comment by shagunsodhani** [comment]
**Not ready for review**

**Comment by shagunsodhani** [comment]
Screen Shot 2020-07-16 at 10 58 38 AM

Attaching screenshot for reference

**Comment by shagunsodhani** [comment]
** Ready for review **

**Comment by omry** [comment]
Looks good!

**Pull request merged**